beacon-chain/operations/voluntaryexits/service.go

```diff
@@ -57,60 +58,51 @@ func (p *Pool) PendingExits(state *beaconstate.BeaconState, slot uint64, noLimit
 }
 
 // InsertVoluntaryExit into the pool. This method is a no-op if the pending exit already exists,
-// has been included recently, or the validator is already exited.
+// or the validator is already exited.
 func (p *Pool) InsertVoluntaryExit(ctx context.Context, state *beaconstate.BeaconState, exit *ethpb.SignedVoluntaryExit) {
 	ctx, span := trace.StartSpan(ctx, "exitPool.InsertVoluntaryExit")
 	defer span.End()
-
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	// Has this validator index been included recently?
-	if p.included[exit.Exit.ValidatorIndex] {
+	// Prevent malformed messages from being inserted.
+	if exit == nil {
+		return
+	}
+	if exit.Exit == nil {
 		return
 	}
 
-	// Has the validator been exited already?
-	if v, err := state.ValidatorAtIndexReadOnly(exit.Exit.ValidatorIndex); err != nil || v.ExitEpoch() != params.BeaconConfig().FarFutureEpoch {
+	// Prevent any sort of duplicate exit from being seenExits.
+	if p.seenExits[exit.Exit.ValidatorIndex] {
 		return
 	}
 
-	// Does this validator exist in the list already? Use binary search to find the answer.
-	if found := sort.Search(len(p.pending), func(i int) bool {
-		e := p.pending[i].Exit
-		return e.ValidatorIndex == exit.Exit.ValidatorIndex
-	}); found != len(p.pending) {
-		// If an exit exists with this validator index, prefer one with an earlier exit epoch.
-		if p.pending[found].Exit.Epoch > exit.Exit.Epoch {
-			p.pending[found] = exit
-		}
+	// Has the validator been exited already?
+	if v, err := state.ValidatorAtIndexReadOnly(exit.Exit.ValidatorIndex); err != nil ||
```

Comment:
Hmm we need to also check that the validator exit doesn't exist in the list. The state is fine but the original bug was with our list.